### PR TITLE
AirfRANS: lr=7e-4 + T_max=10 + weight_decay=1e-2

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Compound the winning AirfRANS lr=7e-4 (baseline 0.01841) with stronger weight decay (1e-2, vs default 1e-4). The lr=7e-4 run showed high volatility at cosine LR peaks — spikes to 0.23-0.27 at epochs 26, 28, 38, 41. Stronger weight decay acts as a form of regularization that may dampen these oscillations, allowing deeper and more consistent basin access during the critical phase transition. Weight decay effectively shrinks the magnitude of weights, potentially providing a smoother objective surface at high LR peaks.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --epochs 999 \
  --agent fern --wandb-name "fern/airfrans-lr7e4-wd1e2" \
  --wandb-group "airfrans-lr7e4-compounds"
```

Key change vs baseline: added `--weight-decay 1e-2` (10x stronger than the default 1e-3) to the winning lr=7e-4+T_max=10 configuration.

Report the best `val_primary/surface_mse` and the epoch it was achieved. Note whether the volatility spikes at cosine peaks are reduced compared to the baseline W&B run (3pbxocca).

## Baseline

Current best AirfRANS metrics (PR #2646, emma — lr=7e-4+T_max=10):
- **val_primary/surface_mse:** 0.01841 ← target to beat
- **full_val/surface_mse_p:** 0.0735
- **full_val/volume_mse:** 0.1331
- **W&B run:** 3pbxocca (41 epochs, best at epoch 35 — earlier phase transition)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine_t_max 10 --no-use-ema --enable-fourier --epochs 999`

External target: 0.0043 — 4.3x gap remaining.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Primary Metric

| Metric | Baseline (lr=7e-4, wd=1e-4) | This Run (lr=7e-4, wd=1e-2) | Delta |
|--------|----------------------------|----------------------------|-------|
| val_primary/surface_mse | 0.01841 | 0.0272 (epoch 39) | +47.7% |
| test_primary/surface_mse | — | 0.2267 (epoch 41*) | N/A* |

*Test evaluated at final epoch 41 (large restart spike), not best val epoch. Not representative.

### Best Val Metrics (Epoch 39)

| Metric | Value |
|--------|-------|
| val_primary/surface_mse | 0.0272 |
| full_val/surface_mse_p | 0.1086 |
| full_val/surface_mse_Ux | 2.59e-5 |
| full_val/surface_mse_Uy | 7.54e-6 |
| full_val/surface_mse_nut | 2.77e-5 |
| full_val/volume_mse | 0.1628 |

### Phase Transition Trajectory

| Epoch | val_primary | surface_mse_p | Phase |
|-------|------------|---------------|-------|
| 1 | 0.3214 | 1.276 | Early |
| 9 | 0.2513 | 1.005 | Plateau |
| 13 | 0.2162 | 0.865 | Transition onset |
| 17 | 0.1215 | 0.485 | Deep transition |
| 22 | 0.1024 | 0.409 | Accelerating |
| 26 | 0.0471 | 0.188 | Deep basin |
| 30 | 0.0328 | 0.129 | Near-best |
| 37 | 0.0313 | 0.125 | Continued descent |
| **39** | **0.0272** | **0.109** | **Best** |

### Volatility Comparison (vs baseline lr=7e-4 wd=1e-4)

The baseline (W&B run 3pbxocca) showed spikes to 0.23-0.27 at cosine LR peaks. This run with wd=1e-2:

| Metric | Baseline (wd=1e-4) | This run (wd=1e-2) |
|--------|--------------------|--------------------|
| Typical restart spike | 0.23-0.27 | 0.19-0.25 |
| Consecutive new-best epochs | ~2-3 | **7** (epochs 11-17) |
| Best-through-restart epochs | Rare | Epochs 22, 27, 30 |
| Overall smoothness | Volatile | Notably smoother |

### Experiment Details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --agent fern --wandb-name "fern/airfrans-lr7e4-wd1e2" --wandb-group "airfrans-lr7e4-compounds"`
- **W&B Run:** `ru3yw0tr`
- **Epochs completed:** 41 (hit 30-min timeout)

### What happened

**The weight decay hypothesis is partially validated — convergence is smoother, but the best val doesn't beat baseline within the same epoch budget.**

1. **Smoother convergence confirmed:** The run produced 7 consecutive new-best epochs (11-17) and achieved new bests even on restart epochs (22, 27, 30) — something the baseline rarely does. The strong wd acts as regularization that dampens the restart oscillation amplitude.

2. **Phase transition occurred:** Starting at epoch 13 (0.216) and reaching 0.0272 at epoch 39. The transition shape is qualitatively similar to the baseline but with smaller restart spikes.

3. **Didn't beat baseline:** Best val 0.0272 vs baseline 0.01841 (+47.7%). Despite smoother convergence, the stronger weight decay may be limiting the model's capacity to reach the deepest part of the basin. The weight regularization that prevents spikes also prevents the model from developing the extreme weight configurations needed for the lowest loss.

4. **Still descending:** The trough-to-trough trend was still going down at cutoff (0.0328 → 0.0313 → 0.0272), suggesting more epochs might help — but the rate of descent was slowing, making it unlikely to reach 0.0184 even with 2x more training.

### Suggested follow-ups

1. **wd=5e-3 (halfway between 1e-4 and 1e-2):** May preserve the smoothing benefits while allowing deeper basin access.
2. **wd decay schedule:** Start with wd=1e-2 for stability during early training, then decay to 1e-4 for deeper basin access later.
3. **wd=1e-2 with longer budget:** The trajectory was still descending — a 60-90 min budget could test whether it converges to or beyond 0.0184.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2678 because ADVISOR: Closing — val_primary/surface_mse=0.02716 is 91% worse than the new baseline 0.01419 (PR #2680, grad-clip=1.0). Weight decay alone (without grad-clip) adds regularization cost without gradient stability benefit. The breakthrough finding is that grad-clip=1.0 is essential at lr=7e-4 — it reduces spike magnitude 40-45% and enables 2.2x deeper basins at cosine troughs. 

The correct follow-up would be grad-clip=1.0 + WD=1e-2 combined, but that's a different hypothesis. fern, please pick up your next assignment..
